README.md
![Roboflow Inference banner](https://github.com/roboflow/inference/blob/main/banner.png?raw=true)

## 🎬 pip install inference

[Roboflow](https://roboflow.com) Inference is the easiest way to use and deploy computer vision models.
Inference supports running object detection, classification, instance segmentation, and even foundation models (like CLIP and SAM).
You can [train and deploy your own custom model](https://github.com/roboflow/notebooks) or use one of the 50,000+
[fine-tuned models shared by the community](https://universe.roboflow.com).

There are three primary `inference` interfaces:
* A Python-native package (`pip install inference`)
* A self-hosted inference server (`inference server start`)
* A [fully-managed, auto-scaling API](https://docs.roboflow.com).

## 🏃 Getting Started

Get up and running with `inference` on your local machine in 3 minutes.

```sh
pip install inference # or inference-gpu if you have CUDA
```

Setup [your Roboflow Private API Key](https://app.roboflow.com/settings/api)
by exporting a `ROBOFLOW_API_KEY` environment variable or
adding it to a `.env` file.

```sh
export ROBOFLOW_API_KEY=your_key_here
```

Run [an open-source Rock, Paper, Scissors model](https://universe.roboflow.com/roboflow-58fyf/rock-paper-scissors-sxsw)
on your webcam stream:

```python
import inference

inference.Stream(
    source="webcam", # or rtsp stream or camera id
    model="rock-paper-scissors-sxsw/11", # from Universe

    on_prediction=lambda predictions, image: (
        print(predictions) # now hold up your hand: 🪨 📄 ✂️
    )
)
```

Now let's extend the example to use [Supervision](https://roboflow.com/supervision)
to visualize the predictions and display them on screen with OpenCV:

```python
import cv2
import inference
import supervision as sv

annotator = sv.BoxAnnotator()

inference.Stream(
    source="webcam", # or rtsp stream or camera id
    model="rock-paper-scissors-sxsw/11", # from Universe

    output_channel_order="BGR",
    use_main_thread=True, # for opencv display
    
    on_prediction=lambda predictions, image: (
        print(predictions), # now hold up your hand: 🪨 📄 ✂️
        
        cv2.imshow(
            "Prediction", 
            annotator.annotate(
                scene=image, 
                detections=sv.Detections.from_roboflow(predictions)
            )
        ),
        cv2.waitKey(1)
    )
)

```

## 👩‍🏫 More Examples

The [`/examples`](https://github.com/roboflow/inference/blob/main/examples/) directory contains code samples for working with and extending `inference` including using foundation models like CLIP, HTTP and UDP clients, and an insights dashboard, along with community examples (PRs welcome)!

## 🎥 Inference in action

Check out Inference running on a video of a football game:

https://github.com/roboflow/inference/assets/37276661/121ab5f4-5970-4e78-8052-4b40f2eec173

## 💻 Why Inference?

Inference provides a scalable method through which you can manage inferences for your vision projects.

Inference is composed of:

- Thousands of [pre-trained community models](https://universe.roboflow.com) that you can use as a starting point.

- Foundation models like CLIP, SAM, and OCR.

- A tight integration with [Supervision](https://roboflow.com/supervision).

- An HTTP server, so you don’t have to reimplement things like image processing and prediction visualization on every project and you can scale your GPU infrastructure independently of your application code, and access your model from whatever language your app is written in.

- Standardized APIs for computer vision tasks, so switching out the model weights and architecture can be done independently of your application code.

- A model registry, so your code can be independent from your model weights & you don't have to re-build and re-deploy every time you want to iterate on your model weights.

- Active Learning integrations, so you can collect more images of edge cases to improve your dataset & model the more it sees in the wild.

- Seamless interoperability with [Roboflow](https://roboflow.com) for creating datasets, training & deploying custom models.

And more!

### 📌 Use the Inference Server

You can learn more about Roboflow Inference Docker Image build, pull and run in our [documentation](https://inference.roboflow.com/quickstart/docker/).

- Run on x86 CPU:

```bash
docker run --net=host roboflow/roboflow-inference-server-cpu:latest
```

- Run on NVIDIA GPU:

```bash
docker run --network=host --gpus=all roboflow/roboflow-inference-server-gpu:latest
```

<details close>
<summary>👉 more docker run options</summary>

- Run on arm64 CPU:

```bash
docker run -p 9001:9001 roboflow/roboflow-inference-server-arm-cpu:latest
```

- Run on NVIDIA GPU with TensorRT Runtime:

```bash
docker run --network=host --gpus=all roboflow/roboflow-inference-server-trt:latest
```

- Run on NVIDIA Jetson with JetPack `4.x`:

```bash
docker run --privileged --net=host --runtime=nvidia roboflow/roboflow-inference-server-jetson:latest
```

- Run on NVIDIA Jetson with JetPack `5.x`:

```bash
docker run --privileged --net=host --runtime=nvidia roboflow/roboflow-inference-server-jetson-5.1.1:latest
```

</details>

### Extras:

Some functionality requires extra dependencies. These can be installed by specifying the desired extras during installation of Roboflow Inference.
| extra | description |
|:-------|:-------------------------------------------------|
| `clip` | Ability to use the core `CLIP` model (by OpenAI) |
| `gaze` | Ability to use the core `Gaze` model |
| `http` | Ability to run the http interface |
| `sam` | Ability to run the core `Segment Anything` model (by Meta AI) |

**_Note:_** Both CLIP and Segment Anything require pytorch to run. These are included in their respective dependencies however pytorch installs can be highly environment dependent. See the [official pytorch install page](https://pytorch.org/get-started/locally/) for instructions specific to your enviornment.

Example install with CLIP dependencies:

```bash
pip install "inference[clip]"
```

## Inference Client

To consume predictions from inference server in Python you can
use the `inference-sdk` package.

```bash
pip install inference-sdk
```

```python
from inference_sdk import InferenceHTTPClient

image_url = "https://media.roboflow.com/inference/soccer.jpg"

# Replace ROBOFLOW_API_KEY with your Roboflow API Key
client = InferenceHTTPClient(
    api_url="http://localhost:9001", # or https://detect.roboflow.com for Hosted API
    api_key="ROBOFLOW_API_KEY"
)
with client.use_model("soccer-players-5fuqs/1"):
    predictions = client.infer(image_url)

print(predictions)
```

Visit our [documentation](https://inference.roboflow.com/) to discover capabilities of `inference-clients` library.

## Single Image Inference

After installing `inference` via pip, you can run a simple inference
on a single image (vs the video stream example above) by instantiating
a `model` and using the `infer` method (don't forget to setup your
`ROBOFLOW_API_KEY` environment variable or `.env` file):

```python
from inference.models.utils import get_roboflow_model

model = get_roboflow_model(
    model_id="soccer-players-5fuqs/1"
)

# you can also infer on local images by passing a file path,
# a PIL image, or a numpy array
results = model.infer(
  image="https://media.roboflow.com/inference/soccer.jpg",
  confidence=0.5,
  iou_threshold=0.5
)

print(results)
```

## Getting CLIP Embeddings

You can run inference with [OpenAI's CLIP model](https://blog.roboflow.com/openai-clip) using:

```python
from inference.models import Clip

image_url = "https://media.roboflow.com/inference/soccer.jpg"

model = Clip()
embeddings = model.embed_image(image_url)

print(embeddings)
```

## Using SAM

You can run inference with [Meta's Segment Anything model](https://blog.roboflow.com/segment-anything-breakdown/) using:

```python
from inference.models import SegmentAnything

image_url = "https://media.roboflow.com/inference/soccer.jpg"

model = SegmentAnything()
embeddings = model.embed_image(image_url)

print(embeddings)
```

## 🏗️ inference Process

To standardize the inference process throughout all our models, Roboflow Inference has a structure for processing inference requests. The specifics can be found on each model's respective page, but overall it works like this for most models:

<img width="900" alt="inference structure" src="https://github.com/stellasphere/inference/assets/29011058/abf69717-f852-4655-9e6e-dae19fc263dc">

## ✅ Supported Models

### Load from Roboflow

You can use models hosted on Roboflow with the following architectures through Inference:

- YOLOv5 Object Detection
- YOLOv5 Instance Segmentation
- YOLOv8 Object Detection
- YOLOv8 Classification
- YOLOv8 Segmentation
- YOLACT Segmentation
- ViT Classification

### Core Models

Core Models are foundation models and models that have not been fine-tuned on a specific dataset.

The following core models are supported:

1. CLIP
2. L2CS (Gaze Detection)
3. Segment Anything (SAM)

## 📝 License

The Roboflow Inference code is distributed under an [Apache 2.0 license](https://github.com/roboflow/inference/blob/master/LICENSE.md). The models supported by Roboflow Inference have their own licenses. View the licenses for supported models below.

| model                     |                                                                license                                                                |
| :------------------------ | :-----------------------------------------------------------------------------------------------------------------------------------: |
| `inference/models/clip`   |                                        [MIT](https://github.com/openai/CLIP/blob/main/LICENSE)                                        |
| `inference/models/gaze`   | [MIT](https://github.com/Ahmednull/L2CS-Net/blob/main/LICENSE), [Apache 2.0](https://github.com/google/mediapipe/blob/master/LICENSE) |
| `inference/models/sam`    |                         [Apache 2.0](https://github.com/facebookresearch/segment-anything/blob/main/LICENSE)                          |
| `inference/models/vit`    |                         [Apache 2.0](https://github.com/roboflow/inference/main/inference/models/vit/LICENSE)                         |
| `inference/models/yolact` |                                     [MIT](https://github.com/dbolya/yolact/blob/master/README.md)                                     |
| `inference/models/yolov5` |                                 [AGPL-3.0](https://github.com/ultralytics/yolov5/blob/master/LICENSE)                                 |
| `inference/models/yolov7` |                                  [GPL-3.0](https://github.com/WongKinYiu/yolov7/blob/main/README.md)                                  |
| `inference/models/yolov8` |                              [AGPL-3.0](https://github.com/ultralytics/ultralytics/blob/master/LICENSE)                               |

## Inference CLI
We've created a CLI tool with useful commands to make the `inference` usage easier. Check out [docs](./inference_cli/README.md).

## 🚀 Enterprise

With a Roboflow Inference Enterprise License, you can access additional Inference features, including:

- Server cluster deployment
- Device management
- Active learning
- YOLOv5 and YOLOv8 commercial license

To learn more, [contact the Roboflow team](https://roboflow.com/sales).

## 📚 documentation

Visit our [documentation](https://inference.roboflow.com) for usage examples and reference for Roboflow Inference.

## 🏆 contribution

We would love your input to improve Roboflow Inference! Please see our [contributing guide](https://github.com/roboflow/inference/blob/master/CONTRIBUTING.md) to get started. Thank you to all of our contributors! 🙏

## 💻 explore more Roboflow open source projects

| Project                                                           | Description                                                                                                                                            |
| :---------------------------------------------------------------- | :----------------------------------------------------------------------------------------------------------------------------------------------------- |
| [supervision](https://roboflow.com/supervision)                   | General-purpose utilities for use in computer vision projects, from predictions filtering and display to object tracking to model evaluation.          |
| [Autodistill](https://github.com/autodistill/autodistill)         | Automatically label images for use in training computer vision models.                                                                                 |
| [Inference](https://github.com/roboflow/inference) (this project) | An easy-to-use, production-ready inference server for computer vision supporting deployment of many popular model architectures and fine-tuned models. |
| [Notebooks](https://roboflow.com/notebooks)                       | Tutorials for computer vision tasks, from training state-of-the-art models to tracking objects to counting objects in a zone.                          |
| [Collect](https://github.com/roboflow/roboflow-collect)           | Automated, intelligent data collection powered by CLIP.                                                                                                |

<br>

<div align="center">

  <div align="center">
      <a href="https://youtube.com/roboflow">
          <img
            src="https://media.roboflow.com/notebooks/template/icons/purple/youtube.png?ik-sdk-version=javascript-1.4.3&updatedAt=1672949634652"
            width="3%"
          />
      </a>
      <img src="https://raw.githubusercontent.com/ultralytics/assets/main/social/logo-transparent.png" width="3%"/>
      <a href="https://roboflow.com">
          <img
            src="https://media.roboflow.com/notebooks/template/icons/purple/roboflow-app.png?ik-sdk-version=javascript-1.4.3&updatedAt=1672949746649"
            width="3%"
          />
      </a>
      <img src="https://raw.githubusercontent.com/ultralytics/assets/main/social/logo-transparent.png" width="3%"/>
      <a href="https://www.linkedin.com/company/roboflow-ai/">
          <img
            src="https://media.roboflow.com/notebooks/template/icons/purple/linkedin.png?ik-sdk-version=javascript-1.4.3&updatedAt=1672949633691"
            width="3%"
          />
      </a>
      <img src="https://raw.githubusercontent.com/ultralytics/assets/main/social/logo-transparent.png" width="3%"/>
      <a href="https://docs.roboflow.com">
          <img
            src="https://media.roboflow.com/notebooks/template/icons/purple/knowledge.png?ik-sdk-version=javascript-1.4.3&updatedAt=1672949634511"
            width="3%"
          />
      </a>
      <img src="https://raw.githubusercontent.com/ultralytics/assets/main/social/logo-transparent.png" width="3%"/>
      <a href="https://disuss.roboflow.com">
          <img
            src="https://media.roboflow.com/notebooks/template/icons/purple/forum.png?ik-sdk-version=javascript-1.4.3&updatedAt=1672949633584"
            width="3%"
          />
      <img src="https://raw.githubusercontent.com/ultralytics/assets/main/social/logo-transparent.png" width="3%"/>
      <a href="https://blog.roboflow.com">
          <img
            src="https://media.roboflow.com/notebooks/template/icons/purple/blog.png?ik-sdk-version=javascript-1.4.3&updatedAt=1672949633605"
            width="3%"
          />
      </a>
      </a>
  </div>

</div>
  </div>


docs/changelog.md
# Changelog

## v0.8.5

### Summary

Contains bug fixes for configurations that use the LICENSE_SERVER setting.

## v0.8.4

### Summary

Image loading is now multi-threaded for batch requests. This should increase total FPS, especially for batch requests that include large images.

The regression test Github action now runs on a Github actions runner.

The extras require has been fixed for the various distribution packages

## v0.8.2

### Summary

Updated the Jetson images so that the default execution provider is CUDA. TensorRT is now an optional configuration via the environment variable ONNXRUNTIME_EXECUTION_PROVIDERS=TensorrtExecutionProvider. The images are also renamed to:

- roboflow/roboflow-inference-server-jetson-4.5.0
- roboflow/roboflow-inference-server-jetson-4.6.1
- roboflow/roboflow-inference-server-jetson-5.1.1

## v0.8.1

### Summary

- Optional Byte Track for UDP interface
- Updated SAM and CLIP requirements and added README quickstarts
- Bug fix for single channel numpy strings

### Breaking Changes

The output for UDP JSON messages has updated the key class_name to class to match HTTP responses.

## v0.8.0

### Summary

This release includes a bit of an overhaul for the model APIs. As this repository started as an internal tool for hosting inference logic, the APIs were tailored for an HTTP interface. With this release, we have made using inference within your python code much smoother and easier. We also updated imports to be less verbose. See the README and docs for new usage. Additionally, a new interface is provided for consuming video streams, and then broadcasting the results over UDP. This interface is tuned for low latency and is ideal for use cases that need to the most up to date information as possible from a video stream. See https://blog.roboflow.com/udp-inference/ for more details.

### Breaking Changes

The main change was creating new definitions for model infer() functions that now take many keyword arguments instead of a single request argument. To continue inferring using request objects, a new method infer_from_request() is provided.

## v0.7.2

Initial release.

client.py
import base64
import requests
from multiprocessing import Pool
import asyncio
import aiohttp


def encode_bas64(image_path, vips=False):
    with open(image_path, "rb") as image:
        x = image.read()
        if vips:
            return x
        image_string = base64.b64encode(x)

    return image_string.decode("ascii")



def request1(i):
    print(i)
    r = requests.post(
        "http://localhost:8000/infer",
        json={"image": encode_bas64("testim.jpg"), "model": "model"},
    )
    print(r.text)
    print(f"FINISHED {i}")
    if r.status_code == 200:
        return 1
    return 0


def request2(i):
    r = requests.post(
        "http://localhost:8000/infer",
        json={"image": encode_bas64("testim.jpg"), "model": "model2"},
    )
    print(r.text)
    print(f"FINISHED {i}")
    if r.status_code == 200:
        return 1
    return 0

def request3(i):
    print(i)
    print(requests.get("http://localhost:8000/").text)
    print(f"FINISHED {i}")

async def do_request(session, i):
    print(f"Starting {i}")
    async with session.post(
        'http://localhost:8000/infer',
        json={"image": encode_bas64("youtube-19-small.jpg"), "model": "model2"},
    ) as response:
        resp = await response.json()
        print(f"Finished {i}")
        print(resp)
        return resp
    
async def do_old_request(session, i):

    project_id = ""
    model_version = ""
    image_url = ""
    confidence = 0.75
    api_key = "Nw3QZal3hhwHP5npbWmw"

    print(f"Starting {i}")
    infer_payload = {
        "image": [{
            "type": "base64",
            "value": encode_bas64("youtube-19-small.jpg"),
        }]*64,
        "confidence": 0.4,
        "iou_threshold": 0.5,
        "api_key": api_key,
        "model_id": "melee/5"
    }
    async with session.post(
        f'http://localhost:9001/infer/object_detection',
        json=infer_payload,
    ) as response:
        resp = await response.json()
        print(f"Finished {i}")
        print(resp)
        return resp


async def do_request_test(session, i):
    print(f"Starting {i}")
    async with session.get(
        'http://localhost:8000/'
    ) as response:
        resp = await response.json()
        print(f"Finished {i}")
        print(resp)
        return resp

def test_image_decompress_speed():
    import pyvips
    from PIL import Image
    import time
    import io
    im = encode_bas64("youtube-19.jpg", vips=True)
    start = time.time()
    im = pyvips.Image.new_from_buffer(im, "")
    im = Image.fromarray(im.numpy())
    im.draft("RGB", (640, 640))
    im.load()
    im = im.resize((640, 640))
    print(f"VIPS Took {time.time() - start} seconds")

    im = encode_bas64("youtube-19.jpg")
    start = time.time()
    decoded_string = io.BytesIO(base64.b64decode(im))
    im = Image.open(decoded_string)
    im.load()
    im = im.resize((640, 640))
    print(f"Pillow Took {time.time() - start} seconds")

    im = encode_bas64("youtube-19.jpg")
    start = time.time()
    decoded_string = io.BytesIO(base64.b64decode(im))
    im = Image.open(decoded_string)
    im.draft("RGB", im.size)
    im.load()
    print(f"Image size after draft {im.size}")
    im = im.resize((640, 640))
    print(f"Pillow (Draft) Took {time.time() - start} seconds")

    start = time.time()
    im = Image.open("youtube-19.jpg")
    print(im.draft("RGB", im.size))
    print(f"Last size {im.size}")
    im.load()
    print(f"Last size {im.size}")
    im = im.resize((640, 640))
    print(f"Pillow (Draft) (FILE) Took {time.time() - start} seconds")
    print(im)
    
    from PIL import features
    print(features.check_feature('libjpeg_turbo'))

async def main():
    tasks = []
    import time
    start = time.time()
    request_batch_size = 64
    num_requests = 1000 // request_batch_size
    connector = aiohttp.TCPConnector(limit=10000, limit_per_host=100000)
    async with aiohttp.ClientSession(read_timeout=0, connector=connector) as session:
        for i in range(num_requests):
            tasks.append(do_old_request(session, i))
        await asyncio.gather(*tasks)
    total = time.time() - start
    print(f"Total time: {total:.2f} seconds")
    print(f"{num_requests * request_batch_size / total} fps")

if __name__ == "__main__":
    from PIL import Image
    im = Image.open("youtube-19.jpg")
    im = im.resize((640, 640))
    im.save("youtube-19-small.jpg")
    loop = asyncio.get_event_loop()
    loop.run_until_complete(main())
    # test_image_decompress_speed()